spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java

```diff
@@ -16,30 +16,71 @@
 
 package org.springframework.integration.util;
 
-import java.io.Closeable;
 import java.io.IOException;
 import java.util.Iterator;
 import java.util.function.Function;
 
+import org.springframework.lang.Nullable;
+
 /**
  * An {@link Iterator} implementation to convert each item from the target
  * {@link #iterator} to a new object applying the {@link #function} on {@link #next()}.
  *
  * @author Artem Bilan
  * @author Ruslan Stelmachenko
+ * @author Gary Russell
  * @since 4.1
  */
 public class FunctionIterator<T, V> implements CloseableIterator<V> {
 
+	private final Object root;
+
 	private final Iterator<T> iterator;
 
 	private final Function<? super T, ? extends V> function;
 
+	/**
+	 * Construct an instance with the provided iterable and function.
+	 * @param iterable the iterable.
+	 * @param function the function.
+	 * @deprecated - use {@link #FunctionIterator(Object, Iterable, Function)}
+	 */
+	@Deprecated
```

Comment:
Why is that? What is wrong with the regular ctor without a `root`?
Please, see your change into the `XPathMessageSplitter`. Looks like the `root` is not required and there is no reason to enforce end-users to think about the `root` where it isn't necessary.

Plus, let's consider a better Javadocs what is this `root` about here.

And that's why I really would prefer a wrapping to the `CloseableIterator` instead...


Also it's not visible here, but looks like we now can remove `remove()` method: it is a `default` one on the `Iterator`

Reply:
My concern is this:

```java
public class Foo implements Iterator, Closeable {

...

}

Since we have no control over what the user returns, I believe we **have** to capture the root object so we can close it when the iteration is complete.

I deprecated the other constructors to discourage their use.

But I can see a better way now; will rework...